File: pkg/scd/uid_generator.go
package scd

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
)

// UIDGenerator generates unique IDs for SCD entities
type UIDGenerator struct{}

// NewUIDGenerator creates a new UID generator
func NewUIDGenerator() *UIDGenerator {
	return &UIDGenerator{}
}

// GenerateUID generates a unique ID for an entity version
func (g *UIDGenerator) GenerateUID(entityType string) (string, error) {
	// Generate random bytes
	randomBytes := make([]byte, 16)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", err
	}
	
	// Encode to base64 URL-safe string without padding
	encodedID := base64.RawURLEncoding.EncodeToString(randomBytes)
	
	// Format: {entityType}_uid_{random-string}
	return fmt.Sprintf("%s_uid_%s", entityType, encodedID), nil
}

// GenerateEntityID generates a new entity ID
func (g *UIDGenerator) GenerateEntityID(entityType string) (string, error) {
	// Generate random bytes
	randomBytes := make([]byte, 16)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", err
	}
	
	// Encode to base64 URL-safe string without padding
	encodedID := base64.RawURLEncoding.EncodeToString(randomBytes)
	
	// Format: {entityType}_{random-string}
	return fmt.Sprintf("%s_%s", entityType, encodedID), nil
}

File: src/main/java/com/mercor/assignment/scd/domain/core/util/UidGenerator.java
package com.mercor.assignment.scd.domain.core.util;

import org.springframework.stereotype.Component;

import java.nio.ByteBuffer;
import java.util.Base64;
import java.util.UUID;

/**
 * Utility for generating unique IDs for SCD entities
 */
@Component
public class UidGenerator {

    /**
     * Generate a unique ID for an entity version
     *
     * @param entityType the type of entity (e.g., "jobs", "timelog")
     * @return a unique ID string
     */
    public String generateUid(String entityType) {
        // Create a unique ID with entity type prefix
        // Format: {entityType}_uid_{random-string}
        
        UUID uuid = UUID.randomUUID();
        String encodedId = encodeUuid(uuid);
        
        return entityType + "_uid_" + encodedId;
    }
    
    /**
     * Generate a new entity ID
     *
     * @param entityType the type of entity
     * @return a unique entity ID
     */
    public String generateEntityId(String entityType) {
        // Create a unique ID with entity type prefix
        // Format: {entityType}_{random-string}
        
        UUID uuid = UUID.randomUUID();
        String encodedId = encodeUuid(uuid);
        
        return entityType + "_" + encodedId;
    }
    
    /**
     * Encode a UUID into a URL-safe string
     */
    private String encodeUuid(UUID uuid) {
        ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
        bb.putLong(uuid.getMostSignificantBits());
        bb.putLong(uuid.getLeastSignificantBits());
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bb.array());
    }
}